classes/importer.php

```diff
@@ -0,0 +1,364 @@
+
+    /** @var boolean $incompatsessdate checks scandate column for records with scan dates that do not match the session date. */
+    public $incompatsessdate = false;
+
+    /** @var boolean $incompatsesstime checks scantime column for records with scan times that are not within the session time. */
+    public $incompatsesstime = false;
+
+    /** @var boolean $multipledays checks scandate column for multiple days of records. */
+    public $multipledays = false;
+
+    /** @var boolean $scantimeerr triggers if the file is suspected of being affected by DST or the wrong file was uploaded. */
+    public $scantimeerr = false;
+
+    /** @var stdClass $highestgradedstatus stores the highest graded status available to assign students. */
+    public $highestgradedstatus;
+
+    /** @var stdClass $lowestgradedstatus stores the lowest graded status available to assign students. */
+    public $lowestgradedstatus;
+
+    /** @var array $validusers only the enrolled users with the correct capability in this course */
+    private $validusers;
+
+    /** @var string $encoding Encoding to use when reading the csv file. Defaults to utf-8. */
+    private $encoding;
+
+    /** @var string $separator How each bit of information is separated in the file. Defaults to comma separated. */
+    private $separator;
+
+    /**
+     * Constructor
+     *
+     * @param string $importid A unique id for this import
+     * @param assign $assignment The current assignment
+     */
+    public function __construct($importid, mod_attendance_structure $att, $encoding = 'utf-8', $separator = 'comma') {
+        $this->importid = $importid;
+        $this->att = $att;
+        $this->encoding = $encoding;
+        $this->separator = $separator;
+    }
+
+    /**
+     * Parse a csv file and save the content to a temp file.
+     * Should be called before init().
+     *
+     * @param string $csvdata The csv data
+     * @return bool false is a failed import
+     */
+    public function parsecsv($csvdata) {
+        $this->csvreader = new csv_import_reader($this->importid, 'attendance');
+        $this->csvreader->load_csv_content($csvdata, $this->encoding, $this->separator);
+    }
+
+    /**
+     * Initialise the import reader and locate the column indexes.
+     *
+     * @return bool false is a failed import
+     */
+    public function init() {
+        GLOBAL $CFG, $DB, $USER;
+
+        if ($this->csvreader == null) {
+            $this->csvreader = new csv_import_reader($this->importid, 'attendance');
+        }
+        $this->csvreader->init();
+
+        $sessioninfo = $this->att->get_session_info($this->att->pageparams->sessionid);
+
+        $filename = $CFG->tempdir.'/csvimport/'.'attendance'.'/'.$USER->id.'/'.$this->importid;
+        $tempfilename = $CFG->tempdir.'/csvimport/'.'attendance'.'/'.$USER->id.'/'.'temp.csv';
+        $header = ['ID Number', 'Symbology', 'Scan Time', 'Scan Date', 'Current Time', 'Current Date', 'Serial Number'];
```

Comment:
you should look at how the other moodle csv import functions work (like uploadcourse) and make this consistent with the other core csv import tools - it uses strtotime() to convert the content in the passed date field.

Reply:
I assume you are specifically referring to the string array in line 146. This list / string array is intended to go at the top of the scan file that is generated by the OPN-200X barcode scanner. 

Interesting thing about this scanner and its software is that when you pull the barcode data off it, it does not include the headings to each column. Now, the way the csvreader->next() function work is it skips the first row of data, assuming it is a header row, and starts reading csv data from the second row. This means the first row of attendance data that was pulled from the barcode scanner would be skipped. We can't have that. So to fix this issue, I wrote lines 144 to 173 to insert a header row (the string array you see in line 146) at the top of the uploaded attendance file. This code is necessary for the operation of this feature. 

The string array in line 146 is simply that, a string. Hence, why I capitalized each word of each heading that was to be inserted at the top of the uploaded file. The ID Number column in the scanner file contains the university-recognized student ID number that is encoded by a barcode on each student's student ID card. This is not the student's user ID number on moodle. The student's user ID number is different from his/her student ID number. It is assumed that each university would store each student's unique university-recognized student ID under the "ID number" field when creating a student's moodle profile. This is number used by this feature to determine their attendance.

Reply:
the problem is that this becomes very specific to your use-case - for this to be accepted into the core atendance module it needs to be more generic and needs to use normal Moodle identifiers.

I'd suggest you take a look at the upload user and  some of the auth/enrol plugins and how they match users against external data - quite often the admin will be able to choose from username/email/idnumber rather than user idnumber which is an optional field and not all sites will use. we also need the "language" to be the same as moodles normal fields. Make sense?

Reply:
I agree that there is an issue with the specificity of my solution with respect to the fact that I mapped the "ID Number" column in the csv file to the "ID number" field in moodle, which is a very specific optional field that many sites may not use. That is what I am getting from your above response. 

My proposed solution is to create a page that will appear after the user uploads a csv file. This new page will show a preview of the uploaded file and allow the user to define which column of the csv file maps to which field in moodle, sort of how the import grades feature operates. Is this more along the lines of what you're looking for? 

![image](https://user-images.githubusercontent.com/45237038/57867759-b3ccf480-77cf-11e9-82b4-0cdbb6ce8274.png)

Also, could you be more specific when you say "we need the language to be the same as moodle normal fields"?

Reply:
Thanks - that sounds much better.

I mean that the strings related to user fields need to match the same language in core - eg there is nothing in mooodle called "Symbology" and on the user profile page the user id number field is referred to as "ID number" - not "ID Number" - does that make more sense? 

thanks!

Reply:
Although - I'm not quite convinced we need "2" drop downs for the user identity and would be happy with just looking at the column names passed in the first row (idnumber/username/email) and then dealing with that automatically but I haven't looked recently at the UI for the import user tool in core moodle.

Reply:
The purpose of the string $header is to act as a header row since the OPN2001 scanner does not generate its own header row, ie the first row of the file contains attendance data. This was needed because the csv import library function next() assumes the first row of any uploaded csv file is a header row and hence we would lose the first row of attendance data it this was not compensated for before processing the file. $header serves no other purpose than to act as a header row. 

I am thinking of removing the strings in the $header variable and leaving it empty to act as a blank row at the top of the file. This is because other persons may use a different format for the csv file, eg the scantime column might be the first column instead of id number. In this case we would need the 2 drop downs so that the user can identify which column in the file contains the user's identity and then link that to the relevant moodle field for processing.

Reply:
I'd prefer to keep the header row and automatically select the field based on the name in the header field. This also means that in future we could add an automated cli option rather than relying on a web based upload- but I'm happy with either option in the first release of this.